### ISSUE TITLE
NSL with TF 2.5 custom loss

### ISSUE THREAD

**Issue by arita37** [tensorflow/neural-structured-learning#94]
Just checking if this is possible
to wrap a custom loss model
using @tf.function  into NSL.

TF is 2.5

what are the constraints ?

**Comment by chunta-lu** [comment]
Could you clarify where do you want to add the custom loss?
The base model used in either GraphRegularization and AdversarialRegularization can use any loss function, so I assume you want to add a custom loss function for the graph regularization loss.
If this is the case, you can add it to the lib/distances.py similar to jensen_shannon_divergence, and add another option in DistanceType in the configs/configs.py.

**Comment by arita37** [comment]
We have a model
wichi cannot use keras fit
because of custom loss.

We want onadd graph regu on top.

Thanks

**Comment by chunta-lu** [comment]
Could you provide the example you of your model and the custom loss?
It sounds like your base model cannot fit with your custom loss. Could you solve this issue first?
The custom loss function should have the same arguments as the other Keras loss function, e.g.,
```
@tf.function
def custom_loss(y_true, y_pred):
 ...
```

**Comment by arita37** [comment]
Sure.

Technically:
   custom_loss with @tf.function
   custom train step with gradient_tape
       with @tf.function
   Manual epoch loop

Should I plug the logits output
into the graph regularizer low level API ?


Thanks

**Comment by csferng** [comment]
Hi @arita37, it looks like you are using a [custom training loop](https://www.tensorflow.org/guide/keras/writing_a_training_loop_from_scratch). (Please correct me if this is not the case.)

To add graph regularization in a custom training loop, you can call [`nsl.keras.layers.PairwiseDistance`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/keras/layers/PairwiseDistance) to compute the graph regularization term and add it to your loss tensor in the gradient_tape scope. Alternatively, you can also call [`nsl.lib.pairwise_distance_wrapper`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/lib/pairwise_distance_wrapper) or other distance functions to calculate the graph regularization term.

The inputs to the distance layer/function are a batch of "embeddings" of seed and neighbor examples, where the "embedding" can be logits or any intermediate layer's output (which your model exposes).

**Comment by arita37** [comment]
Hello,

Thanks, exactly custom training loop.
Also, do we need to provide
the static graph node,vertex in tsv flat file format too ?
(list of neightbors )


Thanks

**Comment by csferng** [comment]
Both [`nsl.lib.pairwise_distance_wrapper()`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/lib/pairwise_distance_wrapper) and  [`nsl.keras.layers.PairwiseDistance().call()`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/keras/layers/PairwiseDistance) assume the graph is already processed, and require two parameters `sources` and `targets` representing the embeddings/logits of seed and neighbor nodes.

One typical way to preprocess the graph for graph regularization is to augment training examples with their neighbor features before training. The function [`nsl.tool.pack_nbrs`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/tools/pack_nbrs) can help achieve this. This function reads a TSV-formatted graph and two TFRecord files and generates a combined TFRecord file. Then in the modeling code you can use [`nsl.lib.unpack_neighbor_features`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/lib/unpack_neighbor_features) or [`nsl.keras.layers.NeighborFeatures`](https://www.tensorflow.org/neural_structured_learning/api_docs/python/nsl/keras/layers/NeighborFeatures) to separate seed and neighbor features in order to compute the embeddings/logits of them.

Please see [this tutorial](https://www.tensorflow.org/neural_structured_learning/tutorials/graph_keras_lstm_imdb) for a concrete example of this workflow. Note that in this tutorial the wrapper `nsl.keras.GraphRegularization` is used for unpacking neighbor features and computing graph regularization.

**Comment by csferng** [comment]
Closing this issue for now. Please feel free to reopen if you have any follow-up questions.
